### PR TITLE
Remove dockerinit once and for all

### LINKED ISSUE #18355: remove dockerinit
the only thing relying on it is https://github.com/docker/libnetwork/blob/9fb7ba8fa01f0ddce99713799e760223c842fb4b/drivers/bridge/setup_bridgenetfiltering.go#L159

cc @rhatdan I think you wanted this as well

cc @tianon

### PR DESCRIPTION
**Pull request by cyphar**
`dockerinit` has been in Docker since (AFAIK) the very beginning. It was once used by the LXC driver to set up containers. LXC has long since been removed and `dockerinit` has no real use anymore.

This could technically be included in `1.10`, but we should probably wait until after the release (although, it would be nice to remove LXC and the former LXC hack known as `dockerinit` in the same version).

Closes #18355

Signed-off-by: Aleksa Sarai asarai@suse.com

### CODE REVIEW

**Comment by cyphar** [comment]
Okay, looks like everything is kosher (flaky tests aside).

**Comment by cyphar** [comment]
This could technically be included in `1.10`, but we should probably wait until after the release (it would be nice to remove LXC and the former LXC hack known as `dockerinit`).

**Comment by cyphar** [comment]
All the tests pass, so after the `engine-api` stuff is merged, we can merge this.

**Comment by hqhq** [comment]
Glad to see we'll finally remove dockerinit, ping @tianon for reviewing.

File: hack/vendor.sh [hunk 1]
```diff
@@ -22,7 +22,7 @@ clone git github.com/vdemeester/shakers 3c10293ce22b900c27acad7b28656196fcc2f73b
 clone git golang.org/x/net 47990a1ba55743e6ef1affd3a14e5bac8553615d https://github.com/golang/net.git
 clone git github.com/docker/go-units 651fc226e7441360384da338d0fd37f2440ffbe3
 clone git github.com/docker/go-connections v0.1.2
-clone git github.com/docker/engine-api v0.2.2
+clone git github.com/docker/engine-api remove-dockerinit https://github.com/cyphar/engine-api
```
Comment (tianon) [inline hack/vendor.sh]: There's a PR open for this change, right? (On my phone now or I'd check directly)

File: hack/make/.go-autogen [hunk 2]
```diff
@@ -44,9 +41,9 @@ When make binary is run, the Dockerfile prepares the build environment by:
 
 make.sh invokes hack/make/.go-autogen to:
 
- - Run rsrc to create a binary file (autogen/winresources/rsrc.syso) that 
-   contains the manifest and icon. This file is automatically picked up by 
-   'go build', so no post-processing steps are required. The sources for 
+ - Run rsrc to create a binary file (autogen/winresources/rsrc.syso) that
+   contains the manifest and icon. This file is automatically picked up by
+   'go build', so no post-processing steps are required. The sources for
```
Comment (tianon) [inline hack/make/.go-autogen]: :heart:!

**Comment by tianon** [comment]
I love it!

File: hack/vendor.sh (hunk 1 above)
Comment (tianon) [inline hack/vendor.sh]: Aha https://github.com/docker/engine-api/issues/53 :+1:

**Comment by LK4D4** [comment]
Very cool!

File: hack/make/.go-autogen (hunk 2 above)
Comment (cyphar) [inline hack/make/.go-autogen]: Haha, my `.vimrc` auto-removes trailing whitespace on `:w`.

**Comment by cyphar** [comment]
It looks like there's some outstanding issues with switching to `libseccomp-dev` instead of building `libseccomp`. I'll leave those for another PR.

**Comment by tianon** [comment]
Yeah, that's fair IMO. :+1:

**Comment by LK4D4** [comment]
Yeah, I think it's enough for this PR :) Thanks @cyphar

**Comment by cyphar** [comment]
@tianon @LK4D4 docker/engine-api#53 is waiting on this to move out of design review before merging.

**Comment by LK4D4** [comment]
@cyphar @calavera I moved it to code-review

**Comment by estesp** [comment]
LGTM

**Comment by thaJeztah** [comment]
@cyphar (and others here) w.r.t. the vendoring; we recently moved the "vendor" check to a separate check in CI; that allows you to include the changed vendor files as part of your PR, without making the tests fail (just the "vendor" check will fail; making it more apparent that we should not merge)

**Comment by thaJeztah** [comment]
Is this something to be considered? https://github.com/docker/docker/issues/18355#issuecomment-161325606

**Comment by cyphar** [comment]
@thaJeztah So you're saying I shouldn't update `vendor.sh` and just manually patch it?

**Comment by thaJeztah** [comment]
@cyphar yes, I think that works; that we can review design/code changes, after that, the PR in the other repository can be merged, vendor.sh can be updated (and run) to get the "real" vendor changes in.

**Comment by thaJeztah** [comment]
For now, this is fine, but I think that's a workflow that may work well (we're still getting used to the new repositories, so bear with us :smile:)

**Comment by rhatdan** [comment]
@cyphar I was talking about things like hooks.d, which we would like to be able to execute from inside of docker containers.  We were looking at the location of dockerinit to figure out where to install a hook.  This is not a big deal, we will just move our hooks to be installed in /usr/lib/docker/ going forward, which seems to be the new standard for distributions, of where to launch helper applications.

**Comment by cyphar** [comment]
The engine-api stuff has been merged, I've updated the PR.

/cc @tianon @hqhq @thaJeztah @LK4D4

**Comment by LK4D4** [comment]
@cyphar OMG, do you use some special go version or stuff? ping @tianon could you help pls?

**Comment by cyphar** [comment]
@LK4D4 Oh god, it failed `vendor` too. I don't want to know what just happened. I'm re-vendoring on an Arch box.

File: hack/make/.go-autogen (hunk 2 above)
Comment (jhowardmsft) [inline hack/make/.go-autogen]: My bad putting the spaces in originally! :heart:

**Comment by cyphar** [comment]
Dammit `DefaultDockerHost` was broken upstream for `freebsd`.

**Comment by calavera** [comment]
@cyphar I've opened https://github.com/docker/engine-api/pull/63 to revert the PullBehavior temporarily in engine-api. That way we'll be able to move forward with this change.

**Comment by cyphar** [comment]
@calavera Are you sure that's necessary? My vendor patch implements the old behaviour (the tests pass AFAICS), so #16609 can actually implement the proper behaviour separately.

**Comment by icecrime** [comment]
```
10:18:08 ---> Making bundle: validate-vendor (in bundles/1.10.0-dev/validate-vendor)
10:21:00 The result of ./hack/vendor.sh differs
10:21:00 
10:21:00  D vendor/src/github.com/docker/engine-api/client/client_nounix.go
10:21:00  D vendor/src/github.com/docker/engine-api/client/client_unix.go
10:21:00 ?? vendor/src/github.com/docker/engine-api/client/client_darwin.go
10:21:00 ?? vendor/src/github.com/docker/engine-api/client/client_linux.go
10:21:00 ?? vendor/src/github.com/docker/engine-api/client/client_windows.go
``` :sob:

**Comment by jhowardmsft** [comment]
@icecrime Can you re-run CI (not on IRC). I fixed this earlier this afternoon

**Comment by cyphar** [comment]
@icecrime Yeah, that's intentional (it was for testing while I waited for the engine-api stuff to be merged, and apparently we're not meant to commit `hack/vendor.sh` changes for unmerged patches?). I'll fix this up tomorrow.

**Comment by vdemeester** [comment]
@cyphar yep, that's a way to validate that changes in vendor are not done manually / are the same as upstream. It's on a separate build to make the *normal* builds be ok but to warn maintainers that there is still some work to do on the vendoring (like waiting for it to be merged upstream) :stuck_out_tongue:.

**Comment by cyphar** [comment]
I've rebased and updated to use the latest engine-api (hopefully that fixes all of our problems).

/cc @tianon @LK4D4 @estesp @calavera @icecrime @vdemeester 

@LK4D4 @tianon: Unrelated to this PR, it looks like my openSUSE machine is breaking `hack/vendor.sh` (cleanup removes files it shouldn't remove). I'll look into this later.

**Comment by estesp** [comment]
still LGTM

**Comment by runcom** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐯
/ping @docker/core-maintainers :smile_cat:

**Comment by tianon** [comment]
Nice, LGTM

**Comment by icecrime** [comment]
LGTM :heartbeat:

**Comment by LK4D4** [comment]
OMG, congrats all!

**Comment by jfrazelle** [comment]
yayyyyy! thanks @cyphar

**Comment by cyphar** [comment]
w00t w00t!

**Comment by tianon** [comment]
This is so good :heart:

**Comment by vdemeester** [comment]
:tada:

**Pull request merged**